test/src/e2e_vm_tests/mod.rs

```diff
@@ -264,16 +264,21 @@ impl TestContext {
                     harness::compile_and_run_unit_tests(&name, &context.run_config, true).await;
                 *output = out;
 
-                let tested_pkg = result.expect("failed to compile and run unit tests");
-                let failed: Vec<String> = tested_pkg
-                    .tests
+                let tested_pkgs = result.expect("failed to compile and run unit tests");
+                let failed: Vec<String> = tested_pkgs
                     .into_iter()
-                    .enumerate()
-                    .filter_map(|(ix, test)| match test.state {
-                        ProgramState::Revert(code) => {
-                            Some(format!("Test {ix} failed with revert {code}\n"))
-                        }
-                        _ => None,
+                    .filter_map(|tested_pkg| {
+                        tested_pkg
+                            .tests
+                            .into_iter()
+                            .enumerate()
+                            .map(|(ix, test)| match test.state {
+                                ProgramState::Revert(code) => {
+                                    Some(format!("Test {ix} failed with revert {code}\n"))
+                                }
+                                _ => None,
+                            })
+                            .collect()
                     })
                     .collect();
```

Comment:
It looks like we might want to use `flat_map` here instead?

We might also want to include the package name now, otherwise there could be multiple tests with the same index. On that note, we should be able to provide the test name rather than just its index too.

Also, it looks like this only checks for the `Revert` state, but tests can now fail if they were decorated with `should_revert` but don't actually revert - maybe this should be updated?

```rust
let failed: Vec<String> = tested_pkgs
    .into_iter()
    .flat_map(|tested_pkg| {
        tested_pkg
            .tests
            .into_iter()
            .filter(|test| !test.passed())
            .map(|test| format!(
                "{}: Test '{}' failed with state {:?}, expected: {:?}",
                tested_pkg.built.name,
                test.name,
                test.state,
                test.condition,
            ))
    })
    .collect();
```